source/common/network/apple_dns_impl.cc

```diff
@@ -52,58 +52,69 @@ AppleDnsResolverStats AppleDnsResolverImpl::generateAppleDnsResolverStats(Stats:
-    if (error != kDNSServiceErr_NoError) {
-      ENVOY_LOG(warn, "DNS resolver error ({}) in dnsServiceGetAddrInfo for {}", error, dns_name);
-      chargeGetAddrInfoErrorStats(error);
-      return nullptr;
-    }
-
-    // If the query was synchronously resolved in the Apple API call, there is no need to return the
-    // query.
-    if (pending_resolution->synchronously_completed_) {
-      return nullptr;
-    }
-
-    // Otherwise, hook up the query's UDS socket to the event loop to process updates.
-    if (!pending_resolution->dnsServiceRefSockFD()) {
-      ENVOY_LOG(warn, "DNS resolver error in dnsServiceRefSockFD for {}", dns_name);
-      return nullptr;
-    }
-
-    pending_resolution->owned_ = true;
-    return pending_resolution.release();
+
+  ENVOY_LOG(trace, "Performing DNS resolution via Apple APIs");
+  auto pending_resolution =
+      std::make_unique<PendingResolution>(*this, callback, dispatcher_, dns_name);
+
+  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+  if (error != kDNSServiceErr_NoError) {
+    ENVOY_LOG(warn, "DNS resolver error ({}) in dnsServiceGetAddrInfo for {}", error, dns_name);
+    chargeGetAddrInfoErrorStats(error);
+    return {nullptr, false};
+  }
+
+  if (pending_resolution->synchronously_completed_) {
+    return {nullptr, true};
+  }
+
+  // Hook up the query's UDS socket to the event loop to process updates.
+  if (!pending_resolution->dnsServiceRefSockFD()) {
+    ENVOY_LOG(warn, "DNS resolver error in dnsServiceRefSockFD for {}", dns_name);
+    return {nullptr, false};
+  }
+
+  return {std::move(pending_resolution), true};
+}
+
+ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
+                                              DnsLookupFamily dns_lookup_family,
+                                              ResolveCb callback) {
+  auto pending_resolution_and_success = startResolution(dns_name, dns_lookup_family, callback);
+
+  // If we synchronously failed the resolution, trigger a failure callback.
+  if (!pending_resolution_and_success.second) {
+    callback(DnsResolver::ResolutionStatus::Failure, {});
+    return nullptr;
+  }
+
+  // We might have synchronously resolved the query, in which case return nullptr.
+  if (!pending_resolution_and_success.first) {
+    return nullptr;
   }
 
-  ASSERT(address != nullptr);
-  // Finish local, synchronous resolution. This needs to happen outside of the exception block above
-  // as the callback itself can throw.
-  callback(DnsResolver::ResolutionStatus::Success,
-           {DnsResponse(address, std::chrono::seconds(60))});
-  return nullptr;
+  // Otherwise return the active resolution query, giving it ownership over itself so that it can
+  // can clean itself up once it's done.
+  pending_resolution_and_success.first->owned_ = true;
+  return pending_resolution_and_success.first.release();
```

Comment:
I think you could move the owned_ part into startResolution() above, and then just always return pending_resolution_and_success.first.release() after the failure handling above?